From 1754c2b5684d0e953865847e6fb6f03443460b7e Mon Sep 17 00:00:00 2001
From: Yossi Etigin <yosefe@Voltaire.COM>
Date: Mon, 11 Aug 2008 19:35:50 +0300
Subject: [PATCH] ib/core: fix for send multicast group send leave retry

Until now, only if joining a multicast group failed there was a retry
mechanism.
This patch will add a mechanism that will retry to leave a multicast
group before giving up.

Signed-off-by: Ron Livne <ronli@voltaire.com>
Signed-off-by: Yossi Etigin <yosefe@voltaire.com>
---
 drivers/infiniband/core/multicast.c |   10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

Index: ofed_kernel/drivers/infiniband/core/multicast.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/multicast.c
+++ ofed_kernel/drivers/infiniband/core/multicast.c
@@ -106,6 +106,8 @@ struct mcast_group {
 	struct ib_sa_query	*query;
 	int			query_id;
 	u16			pkey_index;
+	u8			leave_state;
+	int			retries;
 };
 
 struct mcast_member {
@@ -350,6 +352,7 @@ static int send_leave(struct mcast_group
 
 	rec = group->rec;
 	rec.join_state = leave_state;
+	group->leave_state = leave_state;
 
 	ret = ib_sa_mcmember_rec_query(&sa_client, port->dev->device,
 				       port->port_num, IB_SA_METHOD_DELETE, &rec,
@@ -542,7 +545,11 @@ static void leave_handler(int status, st
 {
 	struct mcast_group *group = context;
 
-	mcast_work_handler(&group->work);
+	if (status && (group->retries > 0) &&
+	    !send_leave(group, group->leave_state))
+		group->retries--;
+	else
+		mcast_work_handler(&group->work);
 }
 
 static struct mcast_group *acquire_group(struct mcast_port *port,
@@ -565,6 +572,7 @@ static struct mcast_group *acquire_group
 	if (!group)
 		return NULL;
 
+	group->retries = 3;
 	group->port = port;
 	group->rec.mgid = *mgid;
 	group->pkey_index = MCAST_INVALID_PKEY_INDEX;
